### PR TITLE
Unit test DOMFragment

### PR DESCRIPTION
**Pull request by jwmerrill**
`DOMFragment` now has 100% line coverage _except for_ the `.toJQ()` method and the `jQToDOMFragment` method. I didn't bother writing tests for these since they will go away by the end of this project.

TODO
- [x] Add test coverage of class manipulation methods
- [x] Add test coverage of *DirOf methods

### CODE REVIEW

File: src/domFragment.ts [hunk 1]
```diff
@@ -58,20 +56,72 @@ class DOMFragment {
...
+   * this fragment that does not overlap this fragment.
    */
   join(sibling: DOMFragment) {
     if (!this.ends) return sibling;
     if (!sibling.ends) return this;
 
     // Check if sibling is actually a sibling of this span
     let found = false;
-    let current: ChildNode | null = this.ends[R];
+    let current: ChildNode | null = this.ends[R].nextSibling;
```
Comment (jwmerrill) [inline src/domFragment.ts:124]: I made this change to make it so that you can't join a fragment to itself.

File: src/domFragment.ts [hunk 2]
```diff
@@ -389,28 +434,6 @@ class DOMFragment {
...
-
-  /**
-   * Return the last Node of this fragment. May be a a Node that is not
-   * an Element such as a Text or Comment node.
-   *
-   * Asserts fragment is not empty.
-   */
-  lastNode() {
-    pray('Fragment is not empty', this.ends);
-    return this.ends[L];
```
Comment (jwmerrill) [inline src/domFragment.ts:411]: Note, I both moved this and fixed the bug that this should be `end[R]`

File: src/domFragment.ts [hunk 3]
```diff
@@ -639,7 +664,7 @@ class DOMFragment {
   toggleClass(classNames: string, on?: boolean) {
     if (on === true) return this.addClass(classNames);
     if (on === false) return this.removeClass(classNames);
-    for (const className of classNames.split(/s+/)) {
+    for (const className of classNames.split(/\s+/)) {
```
Comment (jwmerrill) [inline src/domFragment.ts:667]: Basically dumb luck that unit tests happened to catch this bug. Nasty subtle one.

File: src/domFragment.ts [hunk 4]
```diff
@@ -267,28 +315,22 @@ class DOMFragment {
    * Asserts that this fragment contains exactly one Element.
    */
   prepend(children: DOMFragment) {
-    if (!children.ends) return this;
-    const el = this.oneElement();
-    el.insertBefore(children.toDocumentFragment(), el.firstChild);
+    children.prependTo(this.oneElement());
```
Comment (jwmerrill) [inline src/domFragment.ts:318]: Slight behavior change here: this now throws if this fragment isn't a single element fragment, even if children is empty. I think that's better behavior.

File: src/domFragment.ts (hunk 1 above)
Comment (jwmerrill) [reply src/domFragment.ts]: Update: it turns out that MathQuill was relying on being able to join fragments that share their inner ends. Switched this behavior back in https://github.com/desmosinc/mathquill/pull/196/commits/6362a065635287a99eec3d1cb2da310ecb132738.

Comment (jwmerrill) [reply src/domFragment.ts]: As a later cleanup, I'd be interested in seeing if I can find and remove the places MQ is joining fragments that aren't disjoint. I think it's confusing to allow `join` to work on fragments that overlap by at most one node.

File: src/domFragment.ts [hunk 5]
```diff
@@ -58,12 +77,65 @@ class DOMFragment {
...
+   *
+   * DOMFragments may be invalidated if any of the nodes they contain
+   * are moved or removed independently of the other nodes they contain.
+   */
+  isValid(): boolean {
+    if (!this.ends) return true;
+    let maybeLast: ChildNode | undefined;
+    this.eachNode((el) => (maybeLast = el));
+    return maybeLast === this.ends[R];
+  }
```
Comment (anandthakker) [inline src/domFragment.ts:92]: Wonder if it's worth flagging in the comment that this method is linear in the length of the fragment.  (Imagining a future scenario where we reintroduce quadratic-time selection or something like it just using this innocuous-sounding method to put a defensive check wrong place.)

File: src/domFragment.ts (hunk 3 above)
Comment (anandthakker) [reply src/domFragment.ts:677]: 😬

File: test/unit/domFragment.test.ts [hunk 6]
```diff
@@ -0,0 +1,1697 @@
...
+    });
+    test('throws when called on a fragment with many nodes', () => {
+      const children = [h('span'), h('span')];
+      const parent = h('span', {}, children);
+      assert.throws(() => domFrag(parent).children().children());
+    });
+    test('returns an empty fragment when called on a fragment holding an element with no children', () => {
+      assert.ok(domFrag(h('span')).children().isEmpty());
+    });
+    test('returns an empty fragment when called on a fragment a Text node', () => {
```
Comment (anandthakker) [inline test/unit/domFragment.test.ts:154]: fragment **consisting of** a Text node

File: test/unit/domFragment.test.ts [hunk 7]
```diff
@@ -0,0 +1,1697 @@
...
+      assert.ok(fragmentsEqual(frag1.join(frag2), expected));
+    });
+  });
+
+  suite('.oneNode()', () => {
+    test('throws for empty fragments', () => {
+      assert.throws(() => domFrag().oneNode());
+    });
+
+    test('for many node fragments', () => {
```
Comment (anandthakker) [inline test/unit/domFragment.test.ts:231]: "throws for..."

File: test/unit/domFragment.test.ts [hunk 8]
```diff
@@ -0,0 +1,1697 @@
...
+        return self.insertBefore(arg);
+      },
+    },
+    {
+      name: '.insDirOf(L, ...)',
+      apply: function (self: DOMFragment, arg: DOMFragment) {
+        return self.insDirOf(L, arg);
+      },
+    },
+  ]) {
```
Comment (anandthakker) [inline test/unit/domFragment.test.ts:474]: Nice

File: test/unit/domFragment.test.ts [hunk 9]
```diff
@@ -0,0 +1,1697 @@
...
+          const targetChildren = [h('span'), h.text('a'), h('span')];
+          assert.equal(nTargetChildren, targetChildren.length);
+          const targetParent = h('span', {}, targetChildren);
+          assert.ok(
+            !domFrag(sourceParent).children().isEmpty(),
+            'source parent starts off with children'
+          );
+          const targetFrag = domFrag(
+            domFrag(targetParent).children().toNodeArray()[i]
+          );
```
Comment (anandthakker) [inline test/unit/domFragment.test.ts:532]: Nit: could this just be `const targetFrag = domFrag(targetChildren[i])`?

File: test/unit/domFragment.test.ts [hunk 10]
```diff
@@ -0,0 +1,1697 @@
...
+          const targetChildren = [h('span'), h.text('a'), h('span')];
+          assert.equal(nTargetChildren, targetChildren.length);
+          const targetParent = h('span', {}, targetChildren);
+          assert.ok(
+            !domFrag(sourceParent).children().isEmpty(),
+            'source parent starts off with children'
+          );
+          const targetNodeArray = domFrag(targetParent)
+            .children()
+            .toNodeArray();
```
Comment (anandthakker) [inline test/unit/domFragment.test.ts:566]: Is this the same as `targetChildren`?

**Review by anandthakker — APPROVED** [review]
Whew.  Excellent tests, unenviable authoring experience 😬

File: src/domFragment.ts (hunk 5 above)
Comment (jwmerrill) [reply src/domFragment.ts:92]: Yes, good call.

Comment (jwmerrill) [reply src/domFragment.ts:96]: Addressed in 1131127a54a7f9b7890

File: test/unit/domFragment.test.ts (hunk 6 above)
Comment (jwmerrill) [reply test/unit/domFragment.test.ts]: Thank you. Addressed in 7b99f41717c2fc7a66.

File: test/unit/domFragment.test.ts (hunk 7 above)
Comment (jwmerrill) [reply test/unit/domFragment.test.ts]: Addressed in 7b99f41717c2fc7a66.

File: test/unit/domFragment.test.ts (hunk 9 above)
Comment (jwmerrill) [reply test/unit/domFragment.test.ts]: Yes, nice simplification. Addressed in 7b99f41717c2fc7a66.

File: test/unit/domFragment.test.ts (hunk 10 above)
Comment (jwmerrill) [reply test/unit/domFragment.test.ts]: Addressed in 7b99f41717c2fc7a66.

**Pull request merged**